### ISSUE TITLE
using average to update batch normalization parameters

### ISSUE THREAD

**Issue by leinxx** [guosheng/refinenet#11]
Hi,  I am digging into your code. You are using the 'average' method to train batch normalization layers. Would you mind share the reason for this? What is the difference between using training method "average" and "gradient"? 

Thanks a lot,

Lei

**Comment by guosheng** [comment]
Hi @leinxx 
The method "average" is used in matconvnet for claculating the mean and variance. In our case, we just use the mean and variance in the imagenet pretrain model, so we didn't use this method for the batch normalisation layers.

**Comment by leinxx** [comment]
Thanks a lot for your timely reply. I found out that later yesterday. 

I am currently trying to make a caffe version of your RefineNet as my code has been build on caffe. I did a comparison between the matlab results using your code and the caffe result. The results of the caffe version are not as sharp  as the result from your matlab version. I have been stuck here for two weeks now. Any help would be appreciated.

This is what I did:
1. Build a caffe model proto file based on your matconvnet model. (dropbox link: https://www.dropbox.com/s/3mtdcge0cn9i9y2/RefineNet-101-fix-BN-2chained-pool.solver?dl=0): lr multiplier is 0.1 for resnet and 1 for refinenet part. BN layers are set to use global status. The 4 inputs to refinenet part are **p_ims1d2_outl[1,2,3,4]**.  Their value are copied from the output of **res5c, res4f, res3d, res2c**.

2. For training, The input images are first randomly scaled to between 0.7 and 1.2 of its original size, then this scaled image is rescaled to 1.2 times of its size. So the input to the network can be of scale between [0.7 * 1.2, 1.2 * 1.2] of the original input image. Minus channel-wise 128 and the result is feed to conv1. Since caffe's image_data_layer uses BGR color channel scheme. My own data layer also uses **BGR** color channel scheme. 

3.  While prediction, resize((input_image - [128,128,128]), 1.2) -> conv1. The network outputs are rescaled to the size of input_image by cubic interpolation.

Best,

Lei

**Comment by guosheng** [comment]
Hi Lei,

You could check the learning rate setting in the training demo file. We use 5e-4 for refinenet and 5e-5 for the pretrained resnet. We also add dropout layers in the resnet middle layer outputs if the output dimension is greater than 512, for which the details can be found in the included  network graph of resnet and the network configuration code.
